### PR TITLE
Update resource content

### PR DESCRIPTION
**Pull request by jkralik**

### CODE REVIEW

**Comment by jkralik** [comment]
?       github.com/go-ocf/resource-aggregate/cmd/resource-aggregate-service     [no test files]
ok      github.com/go-ocf/resource-aggregate/cqrs       0.023s  coverage: 100.0% of statements
?       github.com/go-ocf/resource-aggregate/cqrs/commands      [no test files]
?       github.com/go-ocf/resource-aggregate/cqrs/events        [no test files]
?       github.com/go-ocf/resource-aggregate/cqrs/eventstore/test       [no test files]
?       github.com/go-ocf/resource-aggregate/protobuf   [no test files]
?       github.com/go-ocf/resource-aggregate/protobuf/commands  [no test files]
?       github.com/go-ocf/resource-aggregate/protobuf/events    [no test files]
ok      github.com/go-ocf/resource-aggregate/refImpl    10.547s coverage: 100.0% of statements
ok      github.com/go-ocf/resource-aggregate/service    3.864s  coverage: 74.1% of statements
?       github.com/go-ocf/resource-aggregate/uri        [no test files]

File: service/aggregate.go [hunk 1]
```diff
@@ -185,3 +211,45 @@ func (a *aggregate) HandleNotifyContentChanged(ctx context.Context, req commands
 	resp.AuditContext = &auditContext
 	return
 }
+
+// HandleUpdateResourceContent handles a command UpdateResourceContent
+func (a *aggregate) HandleUpdateResourceContent(ctx context.Context, req commands.UpdateResourceContentRequest) (resp commands.UpdateResourceContentResponse, events []cqrsEvent.Event, statusCode int, err error) {
+	if err = validateUpdateResourceContentCommand(req); err != nil {
+		statusCode = http.StatusBadRequest
```
Comment (ondrejtomcik) [inline service/aggregate.go]: why do you return bad request if user is unauthorized?

File: service/aggregate.go [hunk 2]
```diff
@@ -185,3 +211,45 @@ func (a *aggregate) HandleNotifyContentChanged(ctx context.Context, req commands
...
+		err = fmt.Errorf("invalid update resource content command: %v", err)
```
Comment (ondrejtomcik) [inline service/aggregate.go]: command does not has to be necessary invalid. error during resource content update occured

File: service/aggregate.go [hunk 3]
```diff
@@ -185,3 +211,45 @@ func (a *aggregate) HandleNotifyContentChanged(ctx context.Context, req commands
...
+		return
+	}
+
+	events, err = a.ag.HandleCommand(ctx, req)
+	if err != nil {
+		statusCode = http.StatusBadRequest
+		err = fmt.Errorf("unable to process update resource content command: %v", err)
```
Comment (ondrejtomcik) [inline service/aggregate.go]: why our log messages does not contain correlation id? That's one of the ideas behind correlation ids we introduced.

File: service/aggregate.go [hunk 4]
```diff
@@ -185,3 +211,45 @@ func (a *aggregate) HandleNotifyContentChanged(ctx context.Context, req commands
...
+		return
+	}
+	statusCode = http.StatusOK
+
+	auditContext := cqrsUtils.MakeAuditContext(req.AuthorizationContext, req.CorrelationId)
```
Comment (ondrejtomcik) [inline service/aggregate.go]: Why you're defining variable instead of direct assignment ?

File: service/aggregate.go [hunk 5]
```diff
@@ -185,3 +211,45 @@ func (a *aggregate) HandleNotifyContentChanged(ctx context.Context, req commands
...
+	resp.AuditContext = &auditContext
+	return
+}
+
+// HandleNotifyResourceContentUpdateProcessed handles a command NotifyResourceContentUpdateProcessed
+func (a *aggregate) HandleNotifyResourceContentUpdateProcessed(ctx context.Context, req commands.NotifyResourceContentUpdateProcessedRequest) (resp commands.NotifyResourceContentUpdateProcessedResponse, events []cqrsEvent.Event, statusCode int, err error) {
+	if err = validateNotifyResourceContentUpdateProcessedCommand(req); err != nil {
+		statusCode = http.StatusBadRequest
```
Comment (ondrejtomcik) [inline service/aggregate.go]: Same as in previous command. Too generic response status code.

File: service/httpApi.go [hunk 6]
```diff
@@ -220,6 +222,108 @@ func (r RequestHandler) NotifyContentChangedCommandHandler(ctx *fasthttp.Request
...
 	err = publishEvents(ctx, r.publisher, aggregate.DeviceId(), request.ResourceId, events)
 	if err != nil {
-		log.Errorf("cannot publish events for publish command: %v", err)
+		log.Errorf("cannot publish events for notify content changed command: %v", err)
+	}
+}
+
+// UpdateContentHandler handling command update content via http.
+func (r RequestHandler) UpdateContentHandler(ctx *fasthttp.RequestCtx) {
+	log.Debug("start handle update resource content command")
```
Comment (ondrejtomcik) [inline service/httpApi.go]: this is not very informative. which resource at least should be preset

File: service/httpApi.go [hunk 7]
```diff
@@ -220,6 +222,108 @@ func (r RequestHandler) NotifyContentChangedCommandHandler(ctx *fasthttp.Request
...
+		logAndWriteErrorResponse(fmt.Errorf("cannot read request for update resource content command: %v", err), fasthttp.StatusBadRequest, ctx)
+		return
+	}
+
+	if request.AuthorizationContext == nil {
+		logAndWriteErrorResponse(fmt.Errorf("cannot handle update resource content command: authorization context is required"), http.StatusBadRequest, ctx)
+		return
+	}
+
+	correlationID := uuid.Must(uuid.NewV4()).String()
```
Comment (ondrejtomcik) [inline service/httpApi.go]: this should be created in earlier stage to use it for the logging in whole chain within RA

File: service/aggregate.go (hunk 1 above)
Comment (jkralik) [reply service/aggregate.go]: validateUpdateResourceContentCommand just validate structs-> if it has filled all necessary members

File: service/aggregate.go (hunk 2 above)
Comment (jkralik) [reply service/aggregate.go]: same as previous answer: validateUpdateResourceContentCommand just validate structs-> if it has filled all necessary members

File: service/aggregate.go (hunk 3 above)
Comment (jkralik) [reply service/aggregate.go]: this just send err to upper function - its caller response

File: service/aggregate.go (hunk 4 above)
Comment (jkralik) [reply service/aggregate.go]: in golang you cannot assign pointer of result function

File: service/httpApi.go (hunk 6 above)
Comment (jkralik) [reply service/httpApi.go]: i think we can remove all `start handle`.

File: service/httpApi.go (hunk 7 above)
Comment (jkralik) [reply service/httpApi.go]: we need to refactore kil.log to produce logs with correlationId

File: service/httpApi.go (hunk 6 above)
Comment (ondrejtomcik) [reply service/httpApi.go]: Do it please. :)

Comment (jkralik) [reply service/httpApi.go]: removed

File: Makefile [hunk 8]
```diff
@@ -26,6 +27,7 @@ proto/get:
 
 proto/generate:
 	protoc -I=. -I=$GOPATH/src -I=$GOPATH/src/github.com/gogo/protobuf/protobuf --gogofaster_out=$GOPATH/src *.proto
+	goderive ./...
```
Comment (cizmazia) [inline Makefile]: Please could you exclude all generated files using a `.dockerignore` rule and run the code generation for `proto` and `goderive` as part of the build in `Dockerfile`?

File: cqrs/commands/notifyResourceContentUpdateProcessed.go [hunk 9]
```diff
@@ -0,0 +1,11 @@
...
+
+import "github.com/go-ocf/resource-aggregate/protobuf/commands"
+
+type NotifyResourceContentUpdateProcessedRequest struct {
+	commands.NotifyResourceContentUpdateProcessedRequest
+}
+
+type NotifyResourceContentUpdateProcessedResponse struct {
+	commands.NotifyResourceContentUpdateProcessedResponse
+}
```
Comment (cizmazia) [inline cqrs/commands/notifyResourceContentUpdateProcessed.go]: Discuss embedding, extra long names and why acknowledgements need a request and response.

Comment (cizmazia) [reply cqrs/commands/notifyResourceContentUpdateProcessed.go]: #33

File: Makefile (hunk 8 above)
Comment (cizmazia) [reply Makefile]: #34

**Pull request merged**